### PR TITLE
[CURA-9522] Materials list goes off screen.

### PR DESCRIPTION
**Pull request by Joeydelarago**
Nested popups will always try to take up the same vertical area as it's parent. If the popup is smaller than it's parent it won't go below or above the parent. If it is longer than the parent, the bottom will extend past the parent.

CURA-9522

### CODE REVIEW

File: resources/qml/Menus/MaterialBrandMenu.qml [hunk 1]
```diff
@@ -94,16 +94,21 @@ Cura.MenuItem
 
         x: parent.width - UM.Theme.getSize("default_lining").width
         y: {
-            // Checks if popup is more than halfway down the screen AND further than 400 down (this avoids popup going off the top of screen)
-            // If it is then the popup will push up instead of down
-            // This fixes the popups appearing bellow the bottom of the screen.
+            var popupHeight = materialTypesModel.material_types.count * UM.Theme.getSize("menu").height
```
Comment (jellespijker) [inline resources/qml/Menus/MaterialBrandMenu.qml:97]: Does this take spacing into account between the items?

File: resources/qml/Menus/MaterialBrandMenu.qml [hunk 2]
```diff
@@ -94,16 +94,21 @@ Cura.MenuItem
...
+            var spaceToBottom = materialBrandMenu.parent.height - parent.y // Space from hovered item to bottom of list
 
-            if (materialBrandMenu.parent.height / 2 < parent.y && parent.y > 400) {
-                flipped = true
-                return -UM.Theme.getSize("default_lining").width - height + UM.Theme.getSize("menu").height
+            if (popupHeight < spaceToBottom)
+            {
+                return -UM.Theme.getSize("default_lining").width
```
Comment (jellespijker) [inline resources/qml/Menus/MaterialBrandMenu.qml:102]: using `width` here doesn't make sense even though the values are the same:
```suggestion
                return -UM.Theme.getSize("default_lining").height
```

**Review by jellespijker — CHANGES_REQUESTED** [review]

File: resources/qml/Menus/MaterialBrandMenu.qml (hunk 1 above)
Comment (Joeydelarago) [reply resources/qml/Menus/MaterialBrandMenu.qml:97]: The spacing is set to 0 for the Columns below. I do see the alignment is slightly off though

Comment (Joeydelarago) [reply resources/qml/Menus/MaterialBrandMenu.qml:97]: I adjusted the alignment and now they line up perfectly with their parent.

Comment (Joeydelarago) [reply resources/qml/Menus/MaterialBrandMenu.qml:97]: I adjusted the alignment and now they line up perfectly with their parent.

File: resources/qml/Menus/MaterialBrandMenu.qml [hunk 3]
```diff
@@ -16,6 +16,7 @@ Cura.MenuItem
     id: materialBrandMenu
     overrideShowArrow: true
 
+    property var rootHeight
```
Comment (casperlamboo) [inline resources/qml/Menus/MaterialBrandMenu.qml:19]:
```suggestion
    property int rootHeight
```

File: resources/qml/Menus/MaterialBrandMenu.qml [hunk 4]
```diff
@@ -94,16 +96,21 @@ Cura.MenuItem
...
+            {
+                return -UM.Theme.getSize("default_lining").height
+            }
+            else
+            {
+                // The popup is longer than the distance between the hovered item and the bottom of the item list.
+                // This pushes the popup upwards until the bottom lines up with the parent bottom.
+                // Only when popup is longer than the parent, the popup will flow out below the parent.
+                var topOfParent = parent.y + UM.Theme.getSize("narrow_margin").height
+                return -Math.min(parent.y - (materialBrandMenu.parent.height - popupHeight ), topOfParent)
```
Comment (casperlamboo) [inline resources/qml/Menus/MaterialBrandMenu.qml:112]:
```suggestion
                return -Math.min(parent.y - (materialBrandMenu.parent.height - popupHeight), topOfParent)
```

File: resources/qml/Menus/MaterialMenu.qml [hunk 5]
```diff
@@ -101,6 +101,7 @@ Cura.Menu
         delegate: Cura.MaterialBrandMenu
         {
             materialTypesModel: model
+            rootHeight: materialMenu.height
         }
         onObjectAdded: function(index, object) { materialMenu.insertItem(index + 4, object)}
```
Comment (casperlamboo) [inline resources/qml/Menus/MaterialMenu.qml:106]:
```suggestion
        onObjectAdded: function(index, object) {
                materialMenu.insertItem(index + 4, object)
        }
```

File: resources/qml/Menus/MaterialBrandMenu.qml [hunk 6]
```diff
@@ -243,11 +250,37 @@ Cura.MenuItem
...
+
+                            if (popupHeight < spaceToBottom)
+                            {
+                                return -UM.Theme.getSize("default_lining").width
+                            }
+                            else
+                            {
+                                var yAlignedWithTopOfRootPopup = - materialBrandMenu.y - UM.Theme.getSize("default_lining").height
+
+                                if (popupHeight > rootHeight && popupHeight > menuPopup.height)
```
Comment (casperlamboo) [inline resources/qml/Menus/MaterialBrandMenu.qml:266]: I strongly believe in capturing values as "documenting variables". The comment below explains very well what is happening, but also having the variable name telling exactly what is happening helps in my opinion.
```suggestion
                                const parentHeight = menuPopup.height;
                                if (popupHeight > rootHeight && popupHeight > parentHeight)
```

File: resources/qml/Cura.qml [hunk 7]
```diff
@@ -17,6 +17,9 @@ UM.MainWindow
 {
     id: base
 
+    property var main_window_height: base.height
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply resources/qml/Cura.qml:20]:
```suggestion
```

File: resources/qml/Cura.qml [hunk 8]
```diff
@@ -17,6 +17,9 @@ UM.MainWindow
...
+    readonly property var mainWindow: base
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply resources/qml/Cura.qml:20]: https://github.com/Ultimaker/Cura/pull/13975/commits/93a51b6da9a4fd06af91ff88c85580f73e8c4837

**Pull request merged**